File: generator/validate.py
"""Word-square validation.

A completed grid is *solved* when every row and every column is a valid English
word (all cells filled). This is canonical here and mirrored in TypeScript at
``web/src/validate.ts`` so generator/tests and the client agree.
"""

from __future__ import annotations

from typing import cast

from generator.dictionary import Dictionary

Grid = list[list[str | None]]


def is_complete(grid: Grid) -> bool:
    return all(grid[r][c] is not None for r in range(len(grid)) for c in range(len(grid)))


def _is_dictionary_word(word: str, dictionary: Dictionary) -> bool:
    return dictionary.contains(word) or dictionary.contains(word[::-1])


def is_word_square(grid: Grid, dictionary: Dictionary, min_len: int | None = None) -> bool:
    """True iff every row and column is a dictionary word (and grid is complete)."""
    size = len(grid)
    if any(len(row) != size for row in grid):
        return False
    if not is_complete(grid):
        return False
    for r in range(size):
        row_word = "".join(cast(str, grid[r][c]) for c in range(size) if grid[r][c] != "#")
        if (min_len is not None and len(row_word) < min_len) or not _is_dictionary_word(
            row_word, dictionary
        ):
            return False
    for c in range(size):
        col_word = "".join(cast(str, grid[r][c]) for r in range(size) if grid[r][c] != "#")
        if (min_len is not None and len(col_word) < min_len) or not _is_dictionary_word(
            col_word, dictionary
        ):
            return False
    return True


def invalid_cells(
    grid: Grid,
    dictionary: Dictionary,
    min_len: int,
) -> set[tuple[int, int]]:
    """Cells belonging to a fully-filled row/column that is not a dictionary word.

    Powers the client "Check" button: only complete lines are judged, so partial
    lines are never penalized.
    """
    size = len(grid)
    bad: set[tuple[int, int]] = set()
    for r in range(size):
        cells = [(r, c) for c in range(size) if grid[r][c] != "#"]
        if any(grid[r][c] is None for r, c in cells):
            continue
        word = "".join(cast(str, grid[r][c]) for r, c in cells)
        if len(word) < min_len or not _is_dictionary_word(word, dictionary):
            bad.update(cells)
    for c in range(size):
        cells = [(r, c) for r in range(size) if grid[r][c] != "#"]
        if any(grid[r][c] is None for r, c in cells):
            continue
        word = "".join(cast(str, grid[r][c]) for r, c in cells)
        if len(word) < min_len or not _is_dictionary_word(word, dictionary):
            bad.update(cells)
    return bad

File: web/src/validate.ts
import type { Grid } from "./board";
import { BLOCK } from "./board";

// Win/check rules for the word-square game — mirrored from generator/validate.py.
// A grid is solved when every cell is filled and every row and column is a
// dictionary word.

const cellKey = (r: number, c: number) => `${r}-${c}`;

export function isComplete(size: number, grid: Grid): boolean {
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) {
      if (grid[r][c] == null) return false;
    }
  }
  return true;
}

export function isWordSquare(
  size: number,
  grid: Grid,
  dictionary: Set<string>,
  minLen: number,
): boolean {
  if (!isComplete(size, grid)) return false;
  const valid = (word: string) =>
    word.length >= minLen && (dictionary.has(word) || dictionary.has([...word].reverse().join("")));
  for (let r = 0; r < size; r++) {
    let word = "";
    for (let c = 0; c < size; c++) if (grid[r][c] !== BLOCK) word += grid[r][c];
    if (!valid(word)) return false;
  }
  for (let c = 0; c < size; c++) {
    let word = "";
    for (let r = 0; r < size; r++) if (grid[r][c] !== BLOCK) word += grid[r][c];
    if (!valid(word)) return false;
  }
  return true;
}

export interface ValidLines {
  rows: Set<number>;
  cols: Set<number>;
}

/** Indices of rows/columns that currently form a complete dictionary word. */
export function validLines(
  size: number,
  grid: Grid,
  dictionary: Set<string>,
  minLen: number,
): ValidLines {
  const valid = (word: string) =>
    word.length >= minLen && (dictionary.has(word) || dictionary.has([...word].reverse().join("")));
  const rows = new Set<number>();
  const cols = new Set<number>();
  for (let r = 0; r < size; r++) {
    let word = "";
    let complete = true;
    for (let c = 0; c < size; c++) {
      const letter = grid[r][c];
      if (letter === BLOCK) continue;
      if (letter == null) {
        complete = false;
        break;
      }
      word += letter;
    }
    if (complete && valid(word)) rows.add(r);
  }
  for (let c = 0; c < size; c++) {
    let word = "";
    let complete = true;
    for (let r = 0; r < size; r++) {
      const letter = grid[r][c];
      if (letter === BLOCK) continue;
      if (letter == null) {
        complete = false;
        break;
      }
      word += letter;
    }
    if (complete && valid(word)) cols.add(c);
  }
  return { rows, cols };
}

/** Filled cells that do not match the puzzle solution. */
export function incorrectCells(
  size: number,
  grid: Grid,
  solution: string[],
): Set<string> {
  const wrong = new Set<string>();
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) {
      const letter = grid[r][c];
      if (letter != null && letter !== BLOCK && letter !== solution[r]?.[c]) {
        wrong.add(cellKey(r, c));
      }
    }
  }
  return wrong;
}
